### ISSUE TITLE
No changes in the training after bug fix

### ISSUE THREAD

**Issue by jayantsahewal** [microsoft/LightGBM#2482]
## Environment info

Operating System: Linux

CPU/GPU model: CPU

C++/Python/R version: Python

LightGBM version or commit hash: 2.3.0 (Pypi/pip)

## Error message (No error)
Based on this fix - https://github.com/microsoft/LightGBM/pull/2040/files#diff-c4d8a1b2c7a8733615f069f725543826, I was hoping to see change (better/worse not sure) in the training process but it's exactly the same (with the same seed). This is what I am getting with 2.2.x release and 2.3.0 release. Out of 181 features 18 are categorical features. 4 of the categorical features show up in top 10 features. So, this fix should have changed at least something in my experience. Is this expected? If yes, then what did the issue actually fix? Further, more broadly can someone please explain how categorical features are handled in lightGBM. I have gone through the documentation and I didn't find a coherent answer. Thank you for looking into this. I also have made sure that lgb.__version__ points to 2.3.0. Unfortunately, I cannot talk about the use case and/or share the underlying data to reproduce the issue.

Train: (629804, 181) Eval: (157451, 181) Test: (196814, 181)
Training until validation scores don't improve for 500 rounds
[100]	train's auc: 0.847479	train's binary_logloss: 0.501343	valid's auc: 0.756983	valid's binary_logloss: 0.517463
[200]	train's auc: 0.892549	train's binary_logloss: 0.446909	valid's auc: 0.751598	valid's binary_logloss: 0.468684
[300]	train's auc: 0.921564	train's binary_logloss: 0.40718	valid's auc: 0.746589	valid's binary_logloss: 0.431524
[400]	train's auc: 0.941826	train's binary_logloss: 0.374625	valid's auc: 0.742161	valid's binary_logloss: 0.400417
[500]	train's auc: 0.95637	train's binary_logloss: 0.346123	valid's auc: 0.736357	valid's binary_logloss: 0.373203
[600]	train's auc: 0.967273	train's binary_logloss: 0.320509	valid's auc: 0.730499	valid's binary_logloss: 0.348605
Early stopping, best iteration is:
[106]	train's auc: 0.850898	train's binary_logloss: 0.497315	valid's auc: 0.75725	valid's binary_logloss: 0.513923

```
u_clf = lgb.LGBMClassifier(
    boosting_type='gbdt',
    class_weight='balanced',
    n_estimators=5000,
    learning_rate=0.05,
    num_leaves=32,
    max_depth=16,
)

fit_params = {
    'early_stopping_rounds': 500,
    'eval_metric' : 'auc', 
    'eval_set' : [(X_eval, y_eval), (X_train, y_train)],
    'eval_names': ['valid', 'train'],
    'verbose': 100,
    'feature_name': features_lst,
    'categorical_feature': cat_features_lst,
}

u_clf.fit(X_train, y_train, **fit_params)
```

**Comment by guolinke** [comment]
I think this bug is for the forced_split, not for main training logic.
